### PR TITLE
Add permission check to jump past world border (Fixes #2512)

### LINKED ISSUE #2512: /jump goes through vanilla world border
This is a bug report/feature request.

The /jump command can currently teleport you right through the vanilla world border. Maybe make a detection and/or togglable option for that?

Thanks.

### PR DESCRIPTION
**Pull request by willjallen**
Added permission 'essentials.jump.pastWorldBorder'
Added check if player has permission to /jump past the world border

closes #2512

### CODE REVIEW

**Comment by Xeyame** [comment]
Shouldn't all permissions be fully lowercase, and should we make the "You can not jump outside of the world border" message configurable?, Everything else looks good to me.

File: Essentials/src/com/earth2me/essentials/commands/Commandjump.java [hunk 1]
```diff
@@ -46,7 +47,12 @@ public void run(final Server server, final User user, final String commandLabel,
 
         final Trade charge = new Trade(this.getName(), ess);
         charge.isAffordableFor(user);
-        user.getTeleport().teleport(loc, charge, TeleportCause.COMMAND);
+        // If the user is allowed to jump outside of the vanilla world boundary
+        if(Bukkit.getWorld(user.getWorld().getUID()).getWorldBorder().isInside(loc) || user.isAuthorized("essentials.jump.pastWorldBorder")) {
```
Comment (pop4959) [inline Essentials/src/com/earth2me/essentials/commands/Commandjump.java]: You should add a space between the "if" and "(" to follow the convention of the other code in this file. Also, likewise, conventionally permissions are lower case and short / to the point. Something like "essentials.jump.border" might be better.

File: Essentials/src/com/earth2me/essentials/commands/Commandjump.java [hunk 2]
```diff
@@ -46,7 +47,12 @@ public void run(final Server server, final User user, final String commandLabel,
...
+            user.getTeleport().teleport(loc, charge, TeleportCause.COMMAND);
+        }else {
```
Comment (pop4959) [inline Essentials/src/com/earth2me/essentials/commands/Commandjump.java]: space

File: Essentials/src/com/earth2me/essentials/commands/Commandjump.java [hunk 3]
```diff
@@ -46,7 +47,12 @@ public void run(final Server server, final User user, final String commandLabel,
...
+            user.sendMessage("You can not jump outside of the world border");
```
Comment (pop4959) [inline Essentials/src/com/earth2me/essentials/commands/Commandjump.java]: Messages in Essentials should be configurable. You can look at some other code or pull requests in Essentials to get an idea of how you should accomplish this. It is not too hard, and many servers will want to change the message to something else.

**Review by pop4959 — CHANGES_REQUESTED** [review]

**Comment by drtshock** [comment]
You need to add the tl keys to all the translation files. Just leave the message in English though.

**Comment by mdcfe** [comment]
@drtshock Crowdin will add them to each file once the PR is merged and the translation is approved.

**Comment by drtshock** [comment]
ooo

File: Essentials/src/com/earth2me/essentials/Teleport.java [hunk 4]
```diff
@@ -72,6 +72,18 @@ public void cooldown(boolean check) throws Exception {
         }
     }
 
+    private void border(Location loc, IUser teleportee) throws Exception {
+        // TODO: review: should we check if the world is null? I mean I kind of assume a location always has a world.
+        if (!loc.getWorld().getWorldBorder().isInside(loc)
```
Comment (Ichbinjoe) [inline Essentials/src/com/earth2me/essentials/Teleport.java:77]: Locations are always assumed to have a valid world afaik. If they didn't we would have some other problems.

File: Essentials/src/com/earth2me/essentials/Teleport.java [hunk 5]
```diff
@@ -72,6 +72,18 @@ public void cooldown(boolean check) throws Exception {
...
+                && !teleportee.isAuthorized("essentials.teleport.border")) {
+            if (teleportOwner != teleportee) { // I think this will check if you are using #teleportOther()?
```
Comment (Ichbinjoe) [inline Essentials/src/com/earth2me/essentials/Teleport.java:79]: Yes! Comment needs removed.

**Review by Ichbinjoe — APPROVED** [review]
Conditional approval once comments are removed.

File: Essentials/src/com/earth2me/essentials/Teleport.java (hunk 4 above)
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/Teleport.java:77]:
```suggestion
        if (!loc.getWorld().getWorldBorder().isInside(loc)
```

File: Essentials/src/com/earth2me/essentials/Teleport.java (hunk 5 above)
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/Teleport.java:79]:
```suggestion
            if (teleportOwner != teleportee) {
```

**Review by mdcfe — CHANGES_REQUESTED** [review]
A couple of concerns:
* This PR will need to be updated to include AsyncTeleport once it has been merged - see #3102.
* What behaviour exactly will this change? Does this break warps or jails?

**Comment by pop4959** [comment]
Closing due to author inactivity. Can re-open if there are signs of life, especially addressing any concerns that have been mentioned.

**Pull request closed without merging**